### ISSUE TITLE
Props spread incorrectly overrides custom definition

### ISSUE THREAD

**Issue by samuelhulla** [withastro/astro#11920]
### Astro Info

```block
Astro                    v4.15.2
Node                     v20.7.0
System                   macOS (arm64)
Package Manager          pnpm
Output                   static
Adapter                  none
Integrations             @astrojs/tailwind
```

### Describe the Bug

With a super simple component definition

```astro
---
// src/components/Broken.astro
import type { HTMLAttributes } from 'astro/types'
import { cn } from '@/lib/style' // pretty much what clsx does

export type Props = HTMLAttributes<'div'>

const props = Astro.props
---

<div {...props} class={cn("bg-lime-300", props.class)}>
  <slot />
</div>
```

> Also as a side-note I know you can use `class:list` directive which has `clsx` built in, but i just found it easiest to demonstrate this behaviour with styles so I just used this as it's from a real-life use-case for a low-code component library i'm authoring, since in my scenario I'm sometimes using more complex styling imports than just the base `clsx` for stuff like variants and whatnot.

Now let's consider a simple page usage

```astro
---
// pages/page.astro
import Broken from '@/components/Broken.astro'
---

<Broken>this works</Broken>
<Broken class="py-4">this doesnt</Broken>
```

The first instance of the component works fine and renders with the background as expected. However the second one with `class="py-4` does not merge as expected, since the `{...props}` spread overrides the later defined `class` definition.

### How to fix this:

In order to fix this behaviour, we'll need to destructure the conflicting attributes i.e.

```astro
---
// src/components/Workaround.astro
import type { HTMLAttributes } from 'astro/types'
import { cn } from '@/lib/style'

export type Props = HTMLAttributes<'div'>

const { class: className, ...props } = Astro.props
---

<div {...props} class={cn("bg-lime-300", className)}>
  <slot />
</div>
```

> Had to rename `class` to `className` in the destructuring, since `class` is a reserved keyword in js.

Now in my page definition

```astro
---
import Workaround from '@/components/Workaround.astro'
---
<Workaround>this works</Workaround>
<Workaround class="py-4">this also works</Workaround>
```

This fixes the issue, where the classes merge succesfully with the `bg-lime-300` and `py-4` since we destructured the attribute `class` and the `{...props}` rest arg no longer overrides it.

### What's the expected result?

I would expected the attribute spread to be compiled / built in a correct order of the spread, same way as javascript objects treat spread operators, or as react treats it's attributes.

```tsx
import { cn } from '@/lib/style'
import type { DetailedHTMLProps, HTMLAttributes } from 'react'

export type Props = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>

export default function WorksOutOfTheBox(props: Props) {
  return (
    <div {...props} className={cn("bg-lime-300", props.className}>{props.children}</div>
  )
}
```

This works out of the box, the `{...props}`  do not override `className` definition, since the `className` definition takes precedence in definition order over the spread operator.

```tsx
<WorksOutOfTheBox>this works</WorksOutOfTheBox>
<WorksOutOfTheBox className="py-4">this also works</WorksOutOfTheBox>
```
Aligning with the same spread operator behaviour as in javascript objects, i.e.

```js
{
  ...obj,
  foo: 'bar'
}
```

Where if `obj` had  a key `foo`, it would be overriden with the value of `bar`, since the key `foo` was manually defined after the spread operator happened.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-uen4d4?file=src%2Fcomponents%2FBroken.astro

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by mayank99** [comment]
@samuelhulla Thank you for opening this issue! I have been meaning to report this since the very early days of Astro, but a combination of procrastination and low spoons means I've been working around it for literal years at this point 😅

I should note that a similar issue with likely the same root cause is that spread props do *not* override the explicit attributes that come before this. This pattern is useful when you want to set some *default* attribute values that should be overridden by spread props. (I use this pattern all the time in JSX)

```astro
<!-- `type` prop passed to the component should override `type="button"` -->
<button type="button" {...Astro.props}>
  <slot />
</button>
```

See [minimal repro](https://stackblitz.com/edit/withastro-astro-e1kfnl?file=src%2Fcomponents%2FBroken.astro,src%2Fcomponents%2FWorkaround.astro,src%2Fpages%2Findex.astro).

Fixing this would be a breaking change, so this should probably go in 5.0.

**Comment by matthewp** [comment]
Ah, so HTML doesn't work how I would expect here. Overridden props do come after previous ones, but because they all appear in the HTML, the browser actual uses the *first* value as the value, not the last, which is what I would expect. Here's a simpler example of what gets rendered: https://codepen.io/matthewp/pen/abggJgy

So fixing this will be a compiler change, unfortunately. I think what we'd have to do is if there is a spread in the attributes, move any static values into the spread function rather than treat them separately, as we currently do.

**Comment by samuelhulla** [comment]
@matthewp  I see. Kind of begs the question wether astro should adhere more closely to the native HTML behaviour or  align it's behaviour more to the js-land of rendering frameworks.

I still think it's bit better to have it aligned more with the latter option, given a lot of Astro users do come from the React/Vue/Svelte/Solid land, but it's ultimately your call guys.

**Comment by daniele-orlando** [comment]
I still can reproduce this bug with Astro `v5.13.2`. It's a very annoying issue requiring an annoying workaround like using an omit function over the spread properties to preserve JSX expected behavior.

```html
<nav
  {...omit(props, 'class')}
  class:list={['DocNavMenu-d505 std-flex std-flex-column std-gap2', props.class]}
>
  <slot/>
</nav>
```

**Comment by samuelhulla** [comment]
Hey @matthewp  I wanted to spark discussion a little bit, how realistic is it  that this something to be included in one of Astro's releases. I've been considering looking at the compiler and given it a shot to try to do this, but this could technically be classified as a breaking change given for some users the props/styling may behave differently if they had conflict props with spread prior. 

So I kind of wanted to check with you how realistic this is to be part of Astro's release cycle as I wanted to verify with you about Astro's policy on these kind of changes, before I dedicate further time fixing this - so it doesn't just get then dismissed with it would be a breaking change not worth risking.

**Comment by matthewp** [comment]
@samuelhulla I think it would be realistic. Depending on when you get it done it could be on Astro 6, otherwise would need to wait until Astro 7.

**Comment by samuelhulla** [comment]
@matthewp  added a pr attempt, was quite difficult, but feel free to check it once you'll have time :-) 

Hope you have great holidays 🎄

**Comment by matthewp** [comment]
@samuelhulla sorry for not getting to this before, plan on doing that this week though.

**Comment by Syntarex** [comment]
It would be AWESOME if this fix could make it into v6. <3

**Comment by matthewp** [comment]
We don't have a consensus on making this change. The reason this doesn't work is because of how HTML works. If you write `<div style="color: red" style="color: blue">Hello world</div>` the color will be *red*, because HTML uses the *first* attribute it encounters, not the last like in most programming languages.

Agree that this isn't intuitive. But we don't think it's worth making a change because:

- Except for a few exceptions, Astro doesn't alter the code you write in templates to work differently. 
- This is a big breaking change that might affect people without them even realizing it.
- You *can* do what you're trying to do, just put your overrides in front instead of in back `class="my-class" {...props}`

Will leave the issue open a while longer to continue discussion. If there's something I've said here that's not correct please let me know.

**Comment by daniele-orlando** [comment]
@matthewp I understand your points. Let me reply to some of them.

> We don't have a consensus on making this change. The reason this doesn't work is because of how HTML works. ... HTML uses the first attribute it encounters...

Well, this is the point. We are talking about the JSX syntax, not the HTML one. And JSX has its own rules as defined and adopted by major frameworks like React and Solid. Astro is using the JSX syntax, not a new one, and it should stick to the well defined rules of JSX, in order to prevent subtle bugs and cognitive hovered to people used to React (millions of developers). They don't expect HTML behavior, but they expect the JSX syntax behavior.

This is JSX, not HTML.

```jsx
<div
  {...props}
  class:list={['my-box', {active: isActive}, props.class]}
>
  <slot/>
</div>
```

> This is a big breaking change that might affect people without them even realizing it.

I know. In fact, this change would be a perfect fit for a major release of Astro (like 6.0?), not current one, which usually carries other breaking changes too. This change should be documented in the major release **Upgrade Guide** as always (https://docs.astro.build/en/guides/upgrade-to/v5/).

**Comment by samuelhulla** [comment]
Hey I wanted to contribute to discussion here as I do think some interesting points were raised. I'll try to be as objective as possible but I'll be the first to admit I will be understandably tiny bit biased as creator of the issue/pr.

## Why we should implement this ✅

### 1. We already are writing astro code closer to JSX than HTML

As @daniele-orlando  very correctly pointed out - we already are ostensively writing jsx and not html in astro already. 

We're already doing stuff like `{...props}` spread which **does not exist in standard HTML syntax.**

> *Edited point title to reflect @matthewp correction below that it's not exactly JSX. Rest of point I think still stands tho*

### 2. JSX is the modern norm 

- JSX is the modern norm for majority of web frameworks in 2026. Even competing product like `svelte` which also have their own compiler and their slogan literally says *"using languages you already know — HTML, CSS and JavaScript. It’s a love letter to web development"* and despite them taking pride in using languages you already know - specifically HTML - they still choose to use the JSX approach that this would implement
- Parity with React, Preact, Vue, Solid, Svelte, Qwik === easier adoption. There's very few people writing raw html these days as even most courses seem to nowadays teach you html in a framework based fashion

### 3. Idiomatic code

I do agree with @matthewp 's valid point that users can as well do `<div a="1" a="2" />` in html code to enforce 1 value as they might be used with the standard html.

Issue comes with object spreads, which do not exist in the world of HTML. They do however exist in JavaScript and JSX.

```tsx
<div {...defaults} explicit="override" />
```

Reads as: "Start with defaults, then apply my overrides." This mirrors:
  - JS object spread: `{ ...defaults, explicit: "override" }`
  - JSX/TSX syntax aligned with all the modern web frameworks
  - CSS cascade: later rules override earlier
  - General "waterfall" mental model in most programming languages where we read top->bottom & left->right and spreading rightmost values overwrites the left

In comparison, The legacy model does in theory mirror the HTML model of "start with overrides, then apply defaults", but that as we have established in point 1 -- there is not object spread in html so it's hard to argue it even mirrors the HTML patterns properly. You aren't coming purely from the world of HTML and being familiar sprad syntax. You would need at least some familiarity with JS or other programming languages and vast majority do adhere to the defaults first, override later pattern.

### 4. We already are extending other major frameworks

Since astro is heavily built with the `client:` directives  in mind which allows developers to write reactive code in a framework they are familiar with. I do find it bit jarring if we're already in a lot of astro projects extending other frameworks that we would be the only big web framework whcih then goes the opposing direction if we are already integrating them to the astro ecosystem

### 5. It is a major version bump

I agree  that breaking changes are undesireable. But if there's ever a place to do it it's in the major version update which do commonly contain breaking changes in any sort of common semver.

## Why we shouldn't  ❌

### 1. Not standard HTML

I agree there is a small portion of users  who might be caught offguard by this change given it does not adhere to standard HTML.

Personally I would argue  as in prior points that we weren't writing standard HTML to begin with, but nonetheless there will undoubtedly be some users who either come purely from HTML or even those who got used to astro being this way and writing it this way if they were using exclusively astro.

### 2. Breaking change

Even though it would be part of major version bump, breaking changes are never good. That much I agree with


## How to fix these issues ❤️‍🩹

I didn't want to be purely pro change and I do agree there are some valid points on the against side. I think there are two following options, both of which I like that would prevent majority of the damage/pain points from the against arguments.

> Also both of these options do implicitly assume on top of those changes we still would have proper Upgrade Guide as is the norm for all other astro major version as an added benefit

### a) Writing a codemod 

This could allow us to seamlessly port users to next major astro version without them having to worry breaking anything in their legacy codebases. But this is complex and error prone. Personally I'm more in favor of approach in option b)

### b) Adding a compiler setting to astro config

We could add an optional setting to astro config like `jsxPropOrder?: boolean` which would be on my default for astro 6 users and based on user's config we would choose how to process the props order. 

This would allow for users with legacy apps  to seamlessly adopt to jsx syntax order if they wanted and for those who prefer to stay loyal to the HTML ways, they could still choose to turn it off.

## In conclusion 📝

I don't want to be dismissive of those arguing against this change. I do think they raise some valid concerns but all in all to me this seems like the right choice moving forward and becoming aligned with rest of modern web frameworks.

On top of that if we do go the astro config route I do believe this would take care of the majority of issues people might have with this change and would even change it from a breaking change to a opt-in

Let me know what you think and thanks for reading through to the long-winded post. I tried to be as thorough as possible.

**Comment by millette** [comment]
@samuelhulla It's probably too late in the beta stage, but this could happen anytime during v6 lifetime:

1. introduce experimental flag such as `jsxPropOrder` which would default to false to reflect current behavior
2. if set to true, enable new behavior
3. announce it will become the default in v7
4. either move it from experimental to regular config or remove the option completely

We don't have to do it now since it wouldn't be a breaking change, until v7 to give enough time for people to adapt.

**Comment by matthewp** [comment]
@daniele-orlando `.astro` files are not JSX. They superficially look like JSX and were inspired by its syntax but it is very much different. The reason why JSX has this behavior that you prefer is because it compiles attributes to an object literal `{ a: 'a', a: 'b' }` and duplicates follow the rules of JavaScript. Astro components, on the other hand, compile to template literals like:

    html`<div a="b" ${spreadAttributes({ a: 'b' })>`

Which is why this doesn't work today. We would have to intentional treat static parts of attributes as more than just strings. Which we completely could, of course, but it's worth understanding how and why these behaviors are different.

**Comment by samuelhulla** [comment]
Hey @millette 

I agree that does sound like the most reasonable approach going forward. I'll try to give it a go in following days with the config.

Are there any recommendations you would give me to make sure the config to compiler passing/reading is as performant as possible? Or just a standard go file io operation and checking the values?

**Comment by millette** [comment]
@samuelhulla I'm just an interested 3rd party trying to move the conversation along.

**Comment by matthewp** [comment]
We just merged a spec of the compiler here: https://github.com/withastro/compiler/blob/main/SYNTAX_SPEC.md

That repo is the best place to propose changes to how Astro works now. Since this issue is not planned, I'm going to close this one and the associated PR. Thank you for taking the time to work on this nonetheless.
